broker-daemon/orderbook/index.js

```diff
@@ -49,35 +51,45 @@ class Orderbook {
   /**
    * Initialize the orderbook by syncing its state to the Relayer and indexing
    * the orders.
-   * Also includes retry logic if the Relayer fails.
-   * @return {void}
+   * @return {Promise}
    */
   async initialize () {
     this.logger.info(`Initializing market ${this.marketName}...`)
     this.synced = false
 
+    this.logger.debug(`Rebuilding indexes`)
+    await this.index.ensureIndex()
+    await this.askIndex.ensureIndex()
+    await this.bidIndex.ensureIndex()
+
+    await this.watchMarket()
+  }
+
+  /**
+   * Sync orderbook state with the Relayer and retry when it fails
+   * @private
+   * @return {Promise} Resolves when market is being watched (not necessarily when it is synced)
+   */
+  async watchMarket () {
+    this.logger.debug(`Watching market ${this.marketName}...`)
+
     const { baseSymbol, counterSymbol } = this
     const { lastUpdated, sequence } = await this.lastUpdate()
     const params = { baseSymbol, counterSymbol, lastUpdated, sequence }
 
     const watcher = this.relayer.watchMarket(this.eventStore, params)
 
     watcher.on('sync', async () => {
-      this.logger.debug(`Rebuilding indexes`)
-      await this.index.ensureIndex()
-      this.askIndex = await (new AskIndex(this.store)).ensureIndex()
-      this.bidIndex = await (new BidIndex(this.store)).ensureIndex()
-
       this.synced = true
-      this.logger.info(`Market ${this.marketName} initialized.`)
+      this.logger.info(`Market ${this.marketName} synced.`)
     })
 
     watcher.on('end', (error) => {
       this.logger.info(`Market ${this.marketName} unavailable, retrying in ${RETRY_WATCHMARKET}ms`, { error })
 
       // TODO: exponential backoff?
```

Comment:
I added a card for this to make sure we didnt forget to add it once https://github.com/sparkswap/broker/pull/272 is in